### PR TITLE
allow dynamically adding new data to HNSW

### PR DESCRIPTION
**Pull request by robertfeldt**
This is not very pretty and can probably be shortened but I really needed to be able to dynamically add data to a HNSW object and this PR should allow this. I think it might also be useful to others.

I tried to also test that when adding incrementally one would then get exactly the same results as if creating from all the data in a single batch. However, it seems this cannot be guaranteed, even if the same random seed is used, since the order of visiting nodes will be different and thus the connections might not be exactly the same. I'm not sure though so would be great with feedback on if this is actually the right way.

### CODE REVIEW

File: src/interface.jl [hunk 1]
```diff
@@ -69,6 +69,42 @@ add_to_graph!(hnsw::HierarchicalNSW; kwargs...) =
...
+    numnew = length(newdata)
+    startindex = length(hnsw.data) + 1
+    endindex = startindex + numnew - 1
+    extend_added!(hnsw, endindex)
+    extend!(hnsw.lgraph, endindex)
+    extend!(hnsw.vlp, endindex)
+
+    # Now add the new data at end of our existing data
+    for d in newdata
+        push!(hnsw.data, d)
```
Comment (JonasIsensee) [inline src/interface.jl:92]: a single `append!` would probably be better here

**Review by JonasIsensee — COMMENTED** [review]
Thanks for this!

File: src/interface.jl (hunk 1 above)
Comment (robertfeldt) [reply src/interface.jl]: Oh yes, my bad... ;)

Now fixed.

**Comment by robertfeldt** [comment]
Not sure what I can do about this one. It complains about a file missing that I haven't edited or deleted. What am I missing here?

**Comment by JonasIsensee** [comment]
ah, no worries.
I moved the workflow files from another repo and didn't consider that this repo doesn't have a separate documentation page.

**Pull request merged**